dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ExpressionEncoder.java

```diff
@@ -0,0 +1,488 @@
+package com.dat3m.dartagnan.encoding;
+
+import com.dat3m.dartagnan.expression.*;
+import com.dat3m.dartagnan.expression.op.COpBin;
+import com.dat3m.dartagnan.expression.processing.ExpressionVisitor;
+import com.dat3m.dartagnan.program.Register;
+import com.dat3m.dartagnan.program.event.core.Event;
+import com.dat3m.dartagnan.program.memory.MemoryObject;
+import org.sosy_lab.java_smt.api.*;
+import org.sosy_lab.java_smt.api.NumeralFormula.IntegerFormula;
+
+import static com.dat3m.dartagnan.GlobalSettings.getArchPrecision;
+import static java.util.Arrays.asList;
+
+class ExpressionEncoder implements ExpressionVisitor<Formula> {
+
+    private final FormulaManager formulaManager;
+    private final BooleanFormulaManager booleanFormulaManager;
+    private final Event event;
+
+    public ExpressionEncoder(FormulaManager formulaManager, Event event) {
+        this.formulaManager = formulaManager;
+        this.booleanFormulaManager = formulaManager.getBooleanFormulaManager();
+        this.event = event;
+    }
+
+    private IntegerFormulaManager integerFormulaManager() {
+        return formulaManager.getIntegerFormulaManager();
+    }
+
+    private BitvectorFormulaManager bitvectorFormulaManager() {
+        return formulaManager.getBitvectorFormulaManager();
+    }
+
+    public BooleanFormula encodeBoolean(ExprInterface expression) {
```

Comment:
I would rename this to `encodeAsBoolean` to make it more explicit that the expression itself need not be a boolean one.

Same for `encode(As)Integer`.